File: cli/src/checks.ts
import path from 'node:path';
import { mkdir, rm, writeFile } from 'node:fs/promises';
import { randomUUID } from 'node:crypto';
import { spawn } from 'node:child_process';
import YAML from 'yaml';
import TOML from '@iarna/toml';
import { formatLabel, type Format } from './format.js';
import type { Reporter } from './reporter.js';

export type CheckOptions = {
  preserve: string[];
  checkCmd?: string;
  checkCmds?: Record<string, string>;
  reporter?: Reporter;
  label?: string;
  current?: number;
  total?: number;
};

const DEFAULT_PRESERVE = ['code_blocks', 'inline_code', 'urls', 'placeholders'];

export async function validate(
  root: string,
  format: Format,
  output: string,
  source: string,
  options: CheckOptions,
): Promise<void> {
  if (options.reporter && options.label) {
    options.reporter.step('Validating', options.current ?? 0, options.total ?? 0, options.label);
  }

  options.reporter?.log('Checking', `syntax-validator: parse ${formatLabel(format)}`);
  const syntaxError = validateSyntax(format, output, source);
  if (syntaxError) {
    throw new Error(`syntax-validator tool failed: ${syntaxError}`);
  }

  const preserveKinds = resolvePreserve(options.preserve);
  if (preserveKinds.length > 0) {
    options.reporter?.log('Checking', 'preserve-check: verify preserved tokens');
    const preserveError = validatePreserve(output, source, preserveKinds);
    if (preserveError) {
      throw new Error(`preserve-check tool failed: ${preserveError}`);
    }
  }

  const command = selectCheckCommand(format, options.checkCmd, options.checkCmds);
  if (command) {
    options.reporter?.log('Checking', 'custom-command: run check_cmd');
    await runExternal(root, command, output);
  }
}

export function validateSyntax(
  format: Format,
  output: string,
  source?: string,
): string | undefined {
  try {
    switch (format) {
      case 'json': {
        JSON.parse(output);
        return undefined;
      }
      case 'yaml': {
        YAML.parse(output);
        return undefined;
      }
      case 'po': {
        return validatePoThorough(output, source);
      }
      case 'markdown': {
        return validateMarkdown(output);
      }
      case 'text':
        return undefined;
    }
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }

    return String(error);
  }
}

function validateMarkdown(content: string): string | undefined {
  const lines = content.split('\n');
  if (lines.length === 0) {
    return undefined;
  }

  const marker = lines[0]?.trim();
  if (marker !== '---' && marker !== '+++') {
    return undefined;
  }

  let end = -1;
  for (let i = 1; i < lines.length; i += 1) {
    if (lines[i]?.trim() === marker) {
      end = i;
      break;
    }
  }

  if (end < 0) {
    return `markdown frontmatter missing closing ${marker}`;
  }

  const frontmatter = lines.slice(1, end).join('\n');
  try {
    if (marker === '---') {
      YAML.parse(frontmatter);
    } else {
      TOML.parse(frontmatter);
    }
  } catch (error) {
    if (error instanceof Error) {
      return `markdown frontmatter invalid ${marker === '---' ? 'yaml' : 'toml'}: ${error.message}`;
    }

    return `markdown frontmatter invalid ${marker === '---' ? 'yaml' : 'toml'}`;
  }

  return undefined;
}

export function resolvePreserve(kinds: string[]): string[] {
  if (kinds.length === 0) {
    return [...DEFAULT_PRESERVE];
  }

  if (kinds.some((kind) => kind.trim().toLowerCase() === 'none')) {
    return [];
  }

  return kinds.map((kind) => kind.trim().toLowerCase()).filter(Boolean);
}

export function extractPreservables(source: string, preserveKinds: string[]): string[] {
  const set = new Set<string>();
  const output: string[] = [];

  function push(match: string): void {
    if (!set.has(match)) {
      set.add(match);
      output.push(match);
    }
  }

  let working = source;

  if (preserveKinds.includes('code_blocks')) {
    for (const match of source.matchAll(/```[\s\S]*?```/g)) {
      if (match[0]) {
        push(match[0]);
      }
    }
    working = working.replace(/```[\s\S]*?```/g, '');
  }

  if (preserveKinds.includes('inline_code')) {
    for (const match of working.matchAll(/`[^`\n]+`/g)) {
      if (match[0]) {
        push(match[0]);
      }
    }
  }

  if (preserveKinds.includes('urls')) {
    for (const match of working.matchAll(/https?:\/\/[^\s)"'<>]+/g)) {
      if (match[0]) {
        push(match[0]);
      }
    }
  }

  if (preserveKinds.includes('placeholders')) {
    for (const match of working.matchAll(/\{[^\s{}]+\}/g)) {
      if (match[0]) {
        push(match[0]);
      }
    }
  }

  return output;
}

export function validatePreserve(
  output: string,
  source: string,
  preserveKinds: string[],
): string | undefined {
  const tokens = extractPreservables(source, preserveKinds);
  const missing: string[] = [];

  for (const token of tokens) {
    if (!output.includes(token)) {
      missing.push(token);
      if (missing.length >= 5) {
        break;
      }
    }
  }

  if (missing.length > 0) {
    return `preserved tokens missing from output: ${JSON.stringify(missing)}`;
  }

  return undefined;
}

function selectCheckCommand(
  format: Format,
  fallback?: string,
  commands?: Record<string, string>,
): string {
  const formatCmd = commands?.[format];
  if (formatCmd?.trim()) {
    return formatCmd.trim();
  }

  return fallback?.trim() ?? '';
}

export async function runExternal(
  root: string,
  commandTemplate: string,
  content: string,
): Promise<void> {
  if (!root.trim()) {
    throw new Error('external check requires root path');
  }

  const tmpDir = path.join(root, '.glossia', 'tmp');
  await mkdir(tmpDir, { recursive: true });

  const tmpFile = path.join(tmpDir, `check-${Date.now()}-${randomUUID().slice(0, 8)}.tmp`);
  await writeFile(tmpFile, content, 'utf8');

  const command = commandTemplate.replaceAll('{path}', tmpFile);

  try {
    await runShellCommand(command, root);
  } finally {
    await rm(tmpFile, { force: true });
  }
}

async function runShellCommand(command: string, cwd: string): Promise<void> {
  const shell = process.platform === 'win32' ? 'cmd.exe' : 'sh';
  const args = process.platform === 'win32' ? ['/d', '/s', '/c', command] : ['-c', command];

  await new Promise<void>((resolve, reject) => {
    const child = spawn(shell, args, {
      cwd,
      stdio: ['ignore', 'pipe', 'pipe'],
    });

    let stdout = '';
    let stderr = '';

    child.stdout.on('data', (chunk) => {
      stdout += chunk.toString();
    });
    child.stderr.on('data', (chunk) => {
      stderr += chunk.toString();
    });

    child.on('error', (error) => {
      reject(new Error(`external check failed: ${error.message}`));
    });

    child.on('close', (code) => {
      if (code === 0) {
        resolve();
        return;
      }

      const combined = [stderr, stdout].filter(Boolean).join('\n').trim();
      reject(new Error(`external check failed: exit ${code ?? -1}\n${combined}`));
    });
  });
}

type PoEntry = {
  msgid: string;
  msgstr: string;
  hasPlural: boolean;
  pluralMsgstrs: Record<number, string>;
};

function validatePo(content: string): string | undefined {
  const lines = content.split('\n');
  let state = '';
  let hasMsgid = false;
  let hasMsgstr = false;

  for (const rawLine of lines) {
    const line = rawLine.trim();

    if (!line || line.startsWith('#')) {
      continue;
    }

    if (line.startsWith('msgid ')) {
      if (hasMsgid && !hasMsgstr) {
        return 'po entry missing msgstr';
      }
      hasMsgid = true;
      hasMsgstr = false;
      state = 'msgid';
      if (!hasQuotedString(line)) {
        return 'po msgid missing quoted string';
      }
      continue;
    }

    if (line.startsWith('msgid_plural ')) {
      if (state !== 'msgid') {
        return 'po msgid_plural without msgid';
      }
      if (!hasQuotedString(line)) {
        return 'po msgid_plural missing quoted string';
      }
      continue;
    }

    if (line.startsWith('msgstr')) {
      if (!hasMsgid) {
        return 'po msgstr without msgid';
      }
      hasMsgstr = true;
      state = 'msgstr';
      if (!hasQuotedString(line)) {
        return 'po msgstr missing quoted string';
      }
      continue;
    }

    if (line.startsWith('"')) {
      if (!state) {
        return 'po stray quoted string';
      }
      continue;
    }

    return `po invalid line: ${line}`;
  }

  if (hasMsgid && !hasMsgstr) {
    return 'po entry missing msgstr';
  }

  return undefined;
}

export function validatePoThorough(content: string, source?: string): string | undefined {
  const baseError = validatePo(content);
  if (baseError) {
    return baseError;
  }

  const entries = parsePoEntries(content);
  const hasHeader = entries.some((entry) => entry.msgid === '' && entry.msgstr !== '');
  if (!hasHeader && entries.length > 0) {
    return 'po file missing header entry (msgid "" with Content-Type)';
  }

  const headerEntry = entries.find((entry) => entry.msgid === '');
  const pluralCount = headerEntry ? extractPluralFormsCount(headerEntry.msgstr) : undefined;

  if (pluralCount) {
    for (const entry of entries) {
      if (!entry.hasPlural || entry.msgid === '') {
        continue;
      }

      const maxPlural = Math.max(-1, ...Object.keys(entry.pluralMsgstrs).map(Number));
      if (maxPlural + 1 !== pluralCount) {
        return `po plural forms mismatch: header declares nplurals=${pluralCount} but entry for "${truncate(entry.msgid, 40)}" has ${maxPlural + 1} forms`;
      }
    }
  }

  if (source) {
    const sourceEntries = parsePoEntries(source);
    const formatRegex = /%[sdfiu%]|%\([^)]+\)[sdfiu]|\{[0-9]+\}|\{[a-zA-Z_][a-zA-Z0-9_]*\}/g;

    for (const srcEntry of sourceEntries) {
      if (!srcEntry.msgid) {
        continue;
      }

      const translated = entries.find((entry) => entry.msgid === srcEntry.msgid);
      if (!translated || !translated.msgstr) {
        continue;
      }

      const srcFormats = [...srcEntry.msgstr.matchAll(formatRegex)].map((match) => match[0]);
      for (const fmt of srcFormats) {
        if (!translated.msgstr.includes(fmt)) {
          return `po format string "${fmt}" in source msgstr for "${truncate(srcEntry.msgid, 40)}" missing from translation`;
        }
      }
    }
  }

  const untranslated = entries.filter(
    (entry) => !entry.msgid && !entry.msgstr && Object.keys(entry.pluralMsgstrs).length === 0,
  );
  if (untranslated.length > 0) {
    return `po has ${untranslated.length} untranslated entries`;
  }

  return undefined;
}

function parsePoEntries(content: string): PoEntry[] {
  const entries: PoEntry[] = [];

  let msgid = '';
  let msgstr = '';
  let hasPlural = false;
  let pluralMsgstrs: Record<number, string> = {};
  let state = '';
  let pluralIndex: number | undefined;
  let inEntry = false;

  const pushCurrent = (): void => {
    if (!inEntry) {
      return;
    }

    entries.push({
      msgid,
      msgstr,
      hasPlural,
      pluralMsgstrs: { ...pluralMsgstrs },
    });

    msgid = '';
    msgstr = '';
    hasPlural = false;
    pluralMsgstrs = {};
    state = '';
    pluralIndex = undefined;
    inEntry = false;
  };

  for (const rawLine of content.split('\n')) {
    const line = rawLine.trim();

    if (!line || line.startsWith('#')) {
      pushCurrent();
      continue;
    }

    if (line.startsWith('msgid ')) {
      pushCurrent();
      inEntry = true;
      state = 'msgid';
      msgid = extractQuoted(line);
      continue;
    }

    if (line.startsWith('msgid_plural ')) {
      hasPlural = true;
      state = 'msgid_plural';
      continue;
    }

    if (line.startsWith('msgstr[')) {
      const idx = Number.parseInt(line.slice(7), 10);
      pluralIndex = Number.isNaN(idx) ? 0 : idx;
      state = 'msgstr_plural';
      pluralMsgstrs[pluralIndex] = extractQuoted(line);
      continue;
    }

    if (line.startsWith('msgstr ')) {
      state = 'msgstr';
      msgstr = extractQuoted(line);
      continue;
    }

    if (line.startsWith('"')) {
      const continuation = extractQuotedRaw(line);
      if (state === 'msgid') {
        msgid += continuation;
      } else if (state === 'msgstr') {
        msgstr += continuation;
      } else if (state === 'msgstr_plural' && pluralIndex !== undefined) {
        pluralMsgstrs[pluralIndex] = `${pluralMsgstrs[pluralIndex] ?? ''}${continuation}`;
      }
    }
  }

  pushCurrent();
  return entries;
}

function hasQuotedString(line: string): boolean {
  let count = 0;
  let escaped = false;

  for (const char of line) {
    if (char === '\\' && !escaped) {
      escaped = true;
      continue;
    }

    if (char === '"' && !escaped) {
      count += 1;
    }

    escaped = false;
  }

  return count >= 2;
}

function extractQuoted(line: string): string {
  const firstQuote = line.indexOf('"');
  if (firstQuote < 0) {
    return '';
  }

  return extractQuotedRaw(line.slice(firstQuote));
}

function extractQuotedRaw(line: string): string {
  const trimmed = line.trim();
  if (trimmed.length < 2 || !trimmed.startsWith('"') || !trimmed.endsWith('"')) {
    return '';
  }

  return trimmed
    .slice(1, -1)
    .replaceAll('\\n', '\n')
    .replaceAll('\\t', '\t')
    .replaceAll('\\"', '"')
    .replaceAll('\\\\', '\\');
}

function extractPluralFormsCount(header: string): number | undefined {
  const lines = [...header.split('\\n'), ...header.split('\n')];

  for (const line of lines) {
    const normalized = line.trim().toLowerCase();
    if (!normalized.startsWith('plural-forms:')) {
      continue;
    }

    const index = normalized.indexOf('nplurals=');
    if (index < 0) {
      continue;
    }

    const digits = normalized.slice(index + 'nplurals='.length).match(/^\d+/)?.[0];
    if (digits) {
      return Number.parseInt(digits, 10);
    }
  }

  return undefined;
}

function truncate(input: string, max: number): string {
  if (input.length <= max) {
    return input;
  }

  return `${input.slice(0, max)}...`;
}

File: cli/internal/glossia/checks.go
package glossia

import (
	"bytes"
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"runtime"
	"strconv"
	"strings"
	"time"

	"github.com/pelletier/go-toml/v2"
	"gopkg.in/yaml.v3"
)

type CheckOptions struct {
	Preserve  []string
	CheckCmd  string
	CheckCmds map[string]string

	Reporter Reporter
	Label    string
	Current  int
	Total    int
}

var defaultPreserve = []string{"code_blocks", "inline_code", "urls", "placeholders"}

func validate(root string, format Format, output string, source string, options CheckOptions) error {
	if options.Reporter != nil && strings.TrimSpace(options.Label) != "" {
		options.Reporter.Step(verbValidating, options.Current, options.Total, options.Label)
	}

	if options.Reporter != nil {
		options.Reporter.Log(verbChecking, fmt.Sprintf("syntax-validator: parse %s", formatLabel(format)))
	}
	if syntaxErr := validateSyntax(format, output, source); strings.TrimSpace(syntaxErr) != "" {
		return fmt.Errorf("syntax-validator tool failed: %s", syntaxErr)
	}

	preserveKinds := resolvePreserve(options.Preserve)
	if len(preserveKinds) > 0 {
		if options.Reporter != nil {
			options.Reporter.Log(verbChecking, "preserve-check: verify preserved tokens")
		}
		if preserveErr := validatePreserve(output, source, preserveKinds); strings.TrimSpace(preserveErr) != "" {
			return fmt.Errorf("preserve-check tool failed: %s", preserveErr)
		}
	}

	command := selectCheckCommand(format, options.CheckCmd, options.CheckCmds)
	if strings.TrimSpace(command) != "" {
		if options.Reporter != nil {
			options.Reporter.Log(verbChecking, "custom-command: run check_cmd")
		}
		if err := runExternal(root, command, output); err != nil {
			return err
		}
	}

	return nil
}

func validateSyntax(format Format, output string, source string) string {
	switch format {
	case FormatJSON:
		var v any
		if err := json.Unmarshal([]byte(output), &v); err != nil {
			return err.Error()
		}
		return ""
	case FormatYAML:
		var v any
		if err := yaml.Unmarshal([]byte(output), &v); err != nil {
			return err.Error()
		}
		return ""
	case FormatPO:
		return validatePoThorough(output, source)
	case FormatMarkdown:
		return validateMarkdown(output)
	case FormatText:
		return ""
	default:
		return ""
	}
}

func validateMarkdown(content string) string {
	lines := strings.Split(content, "\n")
	if len(lines) == 0 {
		return ""
	}

	marker := strings.TrimSpace(lines[0])
	if marker != "---" && marker != "+++" {
		return ""
	}

	end := -1
	for i := 1; i < len(lines); i++ {
		if strings.TrimSpace(lines[i]) == marker {
			end = i
			break
		}
	}
	if end < 0 {
		return fmt.Sprintf("markdown frontmatter missing closing %s", marker)
	}

	frontmatter := strings.Join(lines[1:end], "\n")
	if marker == "---" {
		var v any
		if err := yaml.Unmarshal([]byte(frontmatter), &v); err != nil {
			return fmt.Sprintf("markdown frontmatter invalid yaml: %s", err.Error())
		}
		return ""
	}

	var v any
	if err := toml.Unmarshal([]byte(frontmatter), &v); err != nil {
		return fmt.Sprintf("markdown frontmatter invalid toml: %s", err.Error())
	}
	return ""
}

func resolvePreserve(kinds []string) []string {
	if len(kinds) == 0 {
		out := make([]string, 0, len(defaultPreserve))
		out = append(out, defaultPreserve...)
		return out
	}

	for _, k := range kinds {
		if strings.ToLower(strings.TrimSpace(k)) == "none" {
			return []string{}
		}
	}

	out := make([]string, 0, len(kinds))
	for _, k := range kinds {
		n := strings.ToLower(strings.TrimSpace(k))
		if n != "" {
			out = append(out, n)
		}
	}
	return out
}

func extractPreservables(source string, preserveKinds []string) []string {
	set := map[string]bool{}
	var output []string

	push := func(match string) {
		if !set[match] {
			set[match] = true
			output = append(output, match)
		}
	}

	working := source

	if containsString(preserveKinds, "code_blocks") {
		var stripped strings.Builder
		for i := 0; i < len(working); {
			start := strings.Index(working[i:], "```")
			if start < 0 {
				stripped.WriteString(working[i:])
				break
			}
			start += i
			stripped.WriteString(working[i:start])

			end := strings.Index(working[start+3:], "```")
			if end < 0 {
				// No closing fence: stop stripping.
				stripped.WriteString(working[start:])
				break
			}
			end = start + 3 + end + 3
			push(working[start:end])
			i = end
		}
		working = stripped.String()
	}

	if containsString(preserveKinds, "inline_code") {
		re := regexp.MustCompile("`[^`\\n]+`")
		for _, match := range re.FindAllString(working, -1) {
			push(match)
		}
	}

	if containsString(preserveKinds, "urls") {
		re := regexp.MustCompile(`https?://[^\s\)"'<>]+`)
		for _, match := range re.FindAllString(working, -1) {
			push(match)
		}
	}

	if containsString(preserveKinds, "placeholders") {
		re := regexp.MustCompile(`\{[^\s{}]+\}`)
		for _, match := range re.FindAllString(working, -1) {
			push(match)
		}
	}

	return output
}

func validatePreserve(output string, source string, preserveKinds []string) string {
	tokens := extractPreservables(source, preserveKinds)
	var missing []string

	for _, token := range tokens {
		if !strings.Contains(output, token) {
			missing = append(missing, token)
			if len(missing) >= 5 {
				break
			}
		}
	}

	if len(missing) > 0 {
		b, _ := json.Marshal(missing)
		return fmt.Sprintf("preserved tokens missing from output: %s", string(b))
	}
	return ""
}

func selectCheckCommand(format Format, fallback string, commands map[string]string) string {
	if strings.TrimSpace(commands[string(format)]) != "" {
		return strings.TrimSpace(commands[string(format)])
	}
	return strings.TrimSpace(fallback)
}

func runExternal(root string, commandTemplate string, content string) error {
	if strings.TrimSpace(root) == "" {
		return fmt.Errorf("external check requires root path")
	}

	tmpDir := filepath.Join(root, ".glossia", "tmp")
	if err := os.MkdirAll(tmpDir, 0o755); err != nil {
		return err
	}

	tmpFile := filepath.Join(tmpDir, fmt.Sprintf("check-%d-%s.tmp", time.Now().UnixMilli(), randHex(8)))
	if err := os.WriteFile(tmpFile, []byte(content), 0o644); err != nil {
		return err
	}

	command := strings.ReplaceAll(commandTemplate, "{path}", tmpFile)

	defer func() {
		_ = os.Remove(tmpFile)
	}()

	if err := runShellCommand(command, root); err != nil {
		return err
	}
	return nil
}

func runShellCommand(command string, cwd string) error {
	var shell string
	var args []string
	if runtime.GOOS == "windows" {
		shell = "cmd.exe"
		args = []string{"/d", "/s", "/c", command}
	} else {
		shell = "sh"
		args = []string{"-c", command}
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Minute)
	defer cancel()

	cmd := exec.CommandContext(ctx, shell, args...)
	cmd.Dir = cwd

	var stdout bytes.Buffer
	var stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	if err := cmd.Start(); err != nil {
		return fmt.Errorf("external check failed: %s", err.Error())
	}

	err := cmd.Wait()
	if err == nil {
		return nil
	}

	exitCode := -1
	var ee *exec.ExitError
	if errors.As(err, &ee) {
		exitCode = ee.ExitCode()
	}

	combined := strings.TrimSpace(strings.Join(filterNonEmpty([]string{strings.TrimSpace(stderr.String()), strings.TrimSpace(stdout.String())}), "\n"))
	return fmt.Errorf("external check failed: exit %d\n%s", exitCode, combined)
}

func filterNonEmpty(parts []string) []string {
	var out []string
	for _, p := range parts {
		if strings.TrimSpace(p) != "" {
			out = append(out, p)
		}
	}
	return out
}

func containsString(list []string, value string) bool {
	for _, item := range list {
		if item == value {
			return true
		}
	}
	return false
}

// --- PO validation ---

type poEntry struct {
	Msgid        string
	Msgstr       string
	HasPlural    bool
	PluralMsgstr map[int]string
}

func validatePo(content string) string {
	lines := strings.Split(content, "\n")
	state := ""
	hasMsgid := false
	hasMsgstr := false

	for _, rawLine := range lines {
		line := strings.TrimSpace(rawLine)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}

		if strings.HasPrefix(line, "msgid ") {
			if hasMsgid && !hasMsgstr {
				return "po entry missing msgstr"
			}
			hasMsgid = true
			hasMsgstr = false
			state = "msgid"
			if !hasQuotedString(line) {
				return "po msgid missing quoted string"
			}
			continue
		}

		if strings.HasPrefix(line, "msgid_plural ") {
			if state != "msgid" {
				return "po msgid_plural without msgid"
			}
			if !hasQuotedString(line) {
				return "po msgid_plural missing quoted string"
			}
			continue
		}

		if strings.HasPrefix(line, "msgstr") {
			if !hasMsgid {
				return "po msgstr without msgid"
			}
			hasMsgstr = true
			state = "msgstr"
			if !hasQuotedString(line) {
				return "po msgstr missing quoted string"
			}
			continue
		}

		if strings.HasPrefix(line, "\"") {
			if state == "" {
				return "po stray quoted string"
			}
			continue
		}

		return fmt.Sprintf("po invalid line: %s", line)
	}

	if hasMsgid && !hasMsgstr {
		return "po entry missing msgstr"
	}
	return ""
}

func validatePoThorough(content string, source string) string {
	if baseErr := validatePo(content); strings.TrimSpace(baseErr) != "" {
		return baseErr
	}

	entries := parsePoEntries(content)
	hasHeader := false
	for _, e := range entries {
		if e.Msgid == "" && e.Msgstr != "" {
			hasHeader = true
			break
		}
	}
	if !hasHeader && len(entries) > 0 {
		return `po file missing header entry (msgid "" with Content-Type)`
	}

	var headerEntry *poEntry
	for i := range entries {
		if entries[i].Msgid == "" {
			headerEntry = &entries[i]
			break
		}
	}
	pluralCount := 0
	if headerEntry != nil {
		if n := extractPluralFormsCount(headerEntry.Msgstr); n > 0 {
			pluralCount = n
		}
	}

	if pluralCount > 0 {
		for _, entry := range entries {
			if !entry.HasPlural || entry.Msgid == "" {
				continue
			}

			maxPlural := -1
			for idx := range entry.PluralMsgstr {
				if idx > maxPlural {
					maxPlural = idx
				}
			}
			if maxPlural+1 != pluralCount {
				return fmt.Sprintf(
					`po plural forms mismatch: header declares nplurals=%d but entry for "%s" has %d forms`,
					pluralCount,
					truncate(entry.Msgid, 40),
					maxPlural+1,
				)
			}
		}
	}

	if strings.TrimSpace(source) != "" {
		sourceEntries := parsePoEntries(source)
		formatRe := regexp.MustCompile(`%[sdfiu%]|%\([^)]+\)[sdfiu]|\{[0-9]+\}|\{[a-zA-Z_][a-zA-Z0-9_]*\}`)

		for _, srcEntry := range sourceEntries {
			if strings.TrimSpace(srcEntry.Msgid) == "" {
				continue
			}

			var translated *poEntry
			for i := range entries {
				if entries[i].Msgid == srcEntry.Msgid {
					translated = &entries[i]
					break
				}
			}
			if translated == nil || strings.TrimSpace(translated.Msgstr) == "" {
				continue
			}

			srcFormats := formatRe.FindAllString(srcEntry.Msgstr, -1)
			for _, fmtStr := range srcFormats {
				if !strings.Contains(translated.Msgstr, fmtStr) {
					return fmt.Sprintf(
						`po format string "%s" in source msgstr for "%s" missing from translation`,
						fmtStr,
						truncate(srcEntry.Msgid, 40),
					)
				}
			}
		}
	}

	untranslated := 0
	for _, entry := range entries {
		if entry.Msgid == "" && entry.Msgstr == "" && len(entry.PluralMsgstr) == 0 {
			untranslated++
		}
	}
	if untranslated > 0 {
		return fmt.Sprintf("po has %d untranslated entries", untranslated)
	}

	return ""
}

func parsePoEntries(content string) []poEntry {
	var entries []poEntry

	msgid := ""
	msgstr := ""
	hasPlural := false
	pluralMsgstrs := map[int]string{}
	state := ""
	var pluralIndex *int
	inEntry := false

	pushCurrent := func() {
		if !inEntry {
			return
		}

		entries = append(entries, poEntry{
			Msgid:        msgid,
			Msgstr:       msgstr,
			HasPlural:    hasPlural,
			PluralMsgstr: copyIntStringMap(pluralMsgstrs),
		})

		msgid = ""
		msgstr = ""
		hasPlural = false
		pluralMsgstrs = map[int]string{}
		state = ""
		pluralIndex = nil
		inEntry = false
	}

	for _, rawLine := range strings.Split(content, "\n") {
		line := strings.TrimSpace(rawLine)

		if line == "" || strings.HasPrefix(line, "#") {
			pushCurrent()
			continue
		}

		if strings.HasPrefix(line, "msgid ") {
			pushCurrent()
			inEntry = true
			state = "msgid"
			msgid = extractQuoted(line)
			continue
		}

		if strings.HasPrefix(line, "msgid_plural ") {
			hasPlural = true
			state = "msgid_plural"
			continue
		}

		if strings.HasPrefix(line, "msgstr[") {
			idxText := line[7:]
			idx, err := strconv.Atoi(idxText)
			if err != nil {
				idx = 0
			}
			pluralIndex = &idx
			state = "msgstr_plural"
			pluralMsgstrs[idx] = extractQuoted(line)
			continue
		}

		if strings.HasPrefix(line, "msgstr ") {
			state = "msgstr"
			msgstr = extractQuoted(line)
			continue
		}

		if strings.HasPrefix(line, "\"") {
			cont := extractQuotedRaw(line)
			if state == "msgid" {
				msgid += cont
			} else if state == "msgstr" {
				msgstr += cont
			} else if state == "msgstr_plural" && pluralIndex != nil {
				pluralMsgstrs[*pluralIndex] = pluralMsgstrs[*pluralIndex] + cont
			}
		}
	}

	pushCurrent()
	return entries
}

func copyIntStringMap(input map[int]string) map[int]string {
	out := map[int]string{}
	for k, v := range input {
		out[k] = v
	}
	return out
}

func hasQuotedString(line string) bool {
	count := 0
	escaped := false
	for _, r := range line {
		if r == '\\' && !escaped {
			escaped = true
			continue
		}
		if r == '"' && !escaped {
			count++
		}
		escaped = false
	}
	return count >= 2
}

func extractQuoted(line string) string {
	first := strings.IndexByte(line, '"')
	if first < 0 {
		return ""
	}
	return extractQuotedRaw(line[first:])
}

func extractQuotedRaw(line string) string {
	trimmed := strings.TrimSpace(line)
	if len(trimmed) < 2 || !strings.HasPrefix(trimmed, "\"") || !strings.HasSuffix(trimmed, "\"") {
		return ""
	}

	out := trimmed[1 : len(trimmed)-1]
	out = strings.ReplaceAll(out, `\n`, "\n")
	out = strings.ReplaceAll(out, `\t`, "\t")
	out = strings.ReplaceAll(out, `\"`, `"`)
	out = strings.ReplaceAll(out, `\\`, `\`)
	return out
}

func extractPluralFormsCount(header string) int {
	lines := append(strings.Split(header, `\n`), strings.Split(header, "\n")...)
	for _, line := range lines {
		normalized := strings.ToLower(strings.TrimSpace(line))
		if !strings.HasPrefix(normalized, "plural-forms:") {
			continue
		}

		idx := strings.Index(normalized, "nplurals=")
		if idx < 0 {
			continue
		}

		rest := normalized[idx+len("nplurals="):]
		digits := ""
		for i := 0; i < len(rest); i++ {
			if rest[i] < '0' || rest[i] > '9' {
				break
			}
			digits += string(rest[i])
		}
		if digits != "" {
			n, err := strconv.Atoi(digits)
			if err == nil {
				return n
			}
		}
	}

	return 0
}

func truncate(input string, max int) string {
	if len(input) <= max {
		return input
	}
	return input[:max] + "..."
}

// --- small helpers ---

func randHex(n int) string {
	if n <= 0 {
		return ""
	}
	b := make([]byte, (n+1)/2)
	if _, err := rand.Read(b); err != nil {
		// Extremely defensive fallback: don't fail checks due to entropy issues.
		for i := range b {
			b[i] = byte(time.Now().UnixNano() >> uint((i%8)*8))
		}
	}

	out := hex.EncodeToString(b)
	if len(out) > n {
		out = out[:n]
	}
	return out
}
